solution/planning_solutions.go

```diff
@@ -67,40 +66,41 @@ func CreatePlanningSolutionCandidate(iter MultiDimIterator, placeClusters [][]ma
 		res.PlaceURLs = append(res.PlaceURLs, place.GetURL())
 	}
 	res.Score = matching.Score(places)
+	res.ID = uuid.NewString()
 	return res, nil
 }
 
-func FindBestPlanningSolutions(candidates []PlanningSolution, topSolutionsCount int64) []PlanningSolution {
+func FindBestPlanningSolutions(placeClusters [][]matching.Place, topSolutionsCount int64, iterator *MultiDimIterator) []PlanningSolution {
 	if topSolutionsCount <= 0 {
 		topSolutionsCount = TopSolutionsCountDefault
 	}
-	m := make(map[string]PlanningSolution) // map for result extraction
-	vertexes := make([]graph.Vertex, len(candidates))
-	for idx, candidate := range candidates {
-		candidateKey := strconv.FormatInt(int64(idx), 10)
-		vertex := graph.Vertex{Name: candidateKey, Key: candidate.Score}
-		vertexes[idx] = vertex
-		m[candidateKey] = candidate
-	}
-	// use limited-size minimum priority queue
+
 	priorityQueue := &graph.MinPriorityQueueVertex{}
-	for _, vertex := range vertexes {
-		if priorityQueue.Len() == int(topSolutionsCount) {
-			top := (*priorityQueue)[0]
-			if vertex.Key > top.Key {
-				heap.Pop(priorityQueue)
+
+	for iterator.HasNext() {
+		var candidate PlanningSolution
+		var err error
+		candidate, err = createPlanningSolutionCandidate(iterator.Status, placeClusters)
+		if err == nil {
+			newVertex := graph.Vertex{Name: candidate.ID, Key: candidate.Score, Object: candidate}
+			if priorityQueue.Len() == int(topSolutionsCount) {
+				topVertex := (*priorityQueue)[0]
+				if topVertex.Key < newVertex.Key {
+					heap.Pop(priorityQueue)
+					heap.Push(priorityQueue, newVertex)
+				}
 			} else {
-				continue
+				heap.Push(priorityQueue, newVertex)
 			}
 		}
-		heap.Push(priorityQueue, vertex)
+		iterator.Next()
 	}
 
 	res := make([]PlanningSolution, 0)
 
 	for priorityQueue.Len() > 0 {
 		top := heap.Pop(priorityQueue).(graph.Vertex)
-		res = append(res, m[top.Name])
+		res = append(res, top.Object.(PlanningSolution))
```

Comment:
I am curious to see these types of new syntax in both line 102 and 103. Something like `x := var.(T)`, T is the data type of variable. Based on my search, does it do two things, create a copy of var in x and check if the data type is T?

Reply:
This syntax converts an `interface{}` to an object of type `T`